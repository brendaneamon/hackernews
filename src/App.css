.page {
  margin: 2rem;
}

.interactions {
  text-align: center;
}

.table {
  margin: 2rem 0;
}

.table-header {
  display: flex;
  line-height: 2.4rem;
  font-size: 1.6rem;
  padding: 0 1rem;
  justify-content: space-between;
}

.table-empty {
  margin: 20rem;
  text-align: center;
  font-size: 1.6rem;
}

.table-row {
  display: flex;
  line-height: 2.4rem;
  white-space: nowrap;
  margin: 1rem 0;
  padding: 1rem;
  background: #fff;
  border: .1rem solid #e3e3e3;
}

.table-header > span {
  overflow: hidden;
  text-overflow: ellipsis;
}

.table-row > span {
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0 .5rem;
}

.button-inline {
  border-width: 0;
  background: transparent;
  color: inherit;
  text-align: inherit;
  -webkit-font-smoothing: inherit;
  padding: 0;
  font-size: inherit;
  cursor: pointer;
}

.button-active {
  border-radius: 0;
  border-bottom: .1rem solid #38bb6c;
}

.centered-content {
  display: flex;
  justify-content: center;
  align-items: center;
}

.flex {
  display: flex;
}

.flex-justify-start {
  justify-content: flex-start;
}

.padding-2 {
  padding: 2rem;
}

.padding-horizontal-1 {
  padding-left: 1rem;
  padding-right: 1rem;
}

.width-10 {
  width: 10%;
}

.width-30 {
  width: 30%;
}

.width-40 {
  width: 40%;
}
