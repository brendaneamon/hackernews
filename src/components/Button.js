import React from 'react'

const Button = ({
  onClick,
  children,
  className = '',
  type = 'button'
}) =>
  <button
    onClick={onClick}
    className={className}
    type={type}
  >
    {children}
  </button>


export default Button
