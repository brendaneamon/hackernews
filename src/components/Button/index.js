import React from 'react'
import PropTypes from 'prop-types'

const Button = ({
  onClick,
  children,
  className,
  type,
}) =>
  <button
    onClick={onClick}
    className={className}
    type={type}
  >
    {children}
  </button>

Button.defaultProps = {
  className: '',
  type: 'button'
}

Button.propTypes = {
  onClick: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  type: PropTypes.string
}

export default Button
