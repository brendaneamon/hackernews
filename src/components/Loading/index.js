import React from 'react'

const Loading = () =>
  <div className="centered-content padding-2">
    Loading . . .
  </div>

export default Loading
