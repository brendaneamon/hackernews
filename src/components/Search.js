import React from 'react'
import Button from './Button'

const Search = ({ value, onChange, onSubmit, children }) =>
  <form>
    {children}
    <input
      type="text"
      value={value}
      onChange={onChange}
    />
    <Button
      onClick={(e) => onSubmit(e)}
      type="submit"
    >
      Submit
    </Button>
  </form>

export default Search
