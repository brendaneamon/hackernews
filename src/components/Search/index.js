import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Button from '../Button'

class Search extends Component {
  componentDidMount() {
    if (this.node) {
      this.node.focus()
    }
  }

  render() {
    const {
      value,
      onChange,
      onSubmit,
      children
    } = this.props

    return (
      <form>
        {children}
        <input
          type="text"
          value={value}
          onChange={onChange}
          ref={(node) => { this.node = node }}
        />
        <Button
          onClick={(e) => onSubmit(e)}
          type="submit"
        >
          Submit
        </Button>
      </form>
    )
  }
}

Search.propTypes = {
  value: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired
}

export default Search
