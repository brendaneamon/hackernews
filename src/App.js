import React, { Component } from 'react'
import Button from './components/Button'
import Search from './components/Search'
import Table from './components/Table'
import { cache, getSearchResults } from './services/api'
import './App.css'

class App extends Component {
  _isMounted = false

  constructor(props) {
    super(props)

    this.state = {
      cacheKey: null,
      error: null,
      result: null,
      searchTerm: 'redux'
    }

    this.onSearchChange = this.onSearchChange.bind(this)
    this.onSearchSubmit = this.onSearchSubmit.bind(this)
    this.onDismiss = this.onDismiss.bind(this)
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this)
  }

  onDismiss(id) {
    const { cacheKey } = this.state
    const { hits, page } = cache.get(cacheKey)

    const result = {
      page,
      hits: hits.filter(item => item.objectID !== id)
    }
    cache.set(cacheKey, result)
    this.setState({ result })
  }

  onSearchChange(event) {
    this.setState({ searchTerm: event.target.value })
  }

  onSearchSubmit(event) {
    const { searchTerm } = this.state
    this.setState({ cacheKey: searchTerm })
    this.fetchSearchTopStories(searchTerm)
    event.preventDefault()
  }

  fetchSearchTopStories(searchTerm, page = 0) {
    getSearchResults(searchTerm, page).then(result => {
      if (this._isMounted) {
        this.setState({ result })
      }
    }).catch(error => {
      if (this._isMounted) {
        this.setState({ error })
        console.error(error)
      }
    })
  }

  componentDidMount() {
    this._isMounted = true
    const { searchTerm } = this.state
    this.setState({ cacheKey: searchTerm })
    this.fetchSearchTopStories(searchTerm)
  }

  componentWillUnmount() {
    this._isMounted = false
  }

  render() {
    const { searchTerm, result, error } = this.state
    let page = 0
    if (result && result.page) {
      page = +result.page
    }

    return (
      <div className="page">
        <div className="interactions">
          <Search
            value={searchTerm}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
          >
            {'Search '}
          </Search>
        </div>
        {error &&
          <p>
            ERROR: problem getting data from Hacker News API
          </p>
        }
        {result &&
          <Table
            list={result.hits}
            onDismiss={this.onDismiss}
          />
        }
        {result &&
          <div className="interactions">
            <Button
              onClick={() => this.fetchSearchTopStories(searchTerm, page + 1)}
            >
              More
            </Button>
          </div>
        }
      </div>
    )
  }
}

export default App
