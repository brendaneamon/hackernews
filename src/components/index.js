import Button from './Button'
import Loading from './Loading'
import Search from './Search'
import SortHeader from './SortHeader'
import Table from './Table'

export {
  Button,
  Loading,
  Search,
  SortHeader,
  Table
}