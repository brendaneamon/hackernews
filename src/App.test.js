import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import App from './App'
import { Search, Button, Table } from './components'

Enzyme.configure({ adapter : new Adapter() })

let component
let div
let element
let tree

describe('App', () => {
  it('renders without crashing', () => {
    div = document.createElement('div')
    ReactDOM.render(<App />, div)
    ReactDOM.unmountComponentAtNode(div)
  })

  it('has a valid snapshot', () => {
    component = renderer.create(<App />)
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('Search', () => {
  const searchProps = {
    onChange() {},
    onSubmit() {}
  }

  it('renders without crashing', () => {
    div = document.createElement('div')
    ReactDOM.render(<Search { ...searchProps }>Search</Search>, div)
    ReactDOM.unmountComponentAtNode(div)
  })

  it('has a valid snapshot', () => {
    component = renderer.create(
      <Search { ...searchProps }>Search</Search>
    )
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('Button', () => {
  const btnProps = {
    className: 'fake-class',
    onClick() {}
  }

  it('renders without crashing', () => {
    div = document.createElement('div')
    ReactDOM.render(<Button { ...btnProps }>Button Text</Button>, div)
    ReactDOM.unmountComponentAtNode(div)
  })

  it('has a valid snapshot', () => {
    component = renderer.create(
      <Button { ...btnProps }>
        Button Text
      </Button>
    )
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('has the expected type', () => {
    element = shallow(
      <Button { ...btnProps }>
        Button Text
      </Button>
    )

    expect(element.text()).toBe('Button Text')
  })
})

describe('Table', () => {
  const tableProps = {
    list: [
      { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'a' },
      { title: '2', author: '2', num_comments: 1, points: 2, objectID: 'b' }
    ]
  }

  it('renders without crashing', () => {
    div = document.createElement('div')
    ReactDOM.render(<Table { ...tableProps } />, div)
    ReactDOM.unmountComponentAtNode(div)
  })

  it('has a valid snapshot', () => {
    component = renderer.create(
      <Table { ...tableProps }/>
    )
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('shows two items in the list', () => {
    element = shallow(
      <Table { ...tableProps }/>
    )
    const list = element.find('.table-row')

    expect(list.length).toBe(2)
  })
})
