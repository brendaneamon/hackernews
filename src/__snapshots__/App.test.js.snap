// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form>
      Search 
      <input
        onChange={[Function]}
        type="text"
        value="redux"
      />
      <button
        className=""
        onClick={[Function]}
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
  <div
    className="interactions"
  >
    <div
      className="centered-content padding-2"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-spinner fa-w-16 fa-pulse fa-3x "
        data-icon="spinner"
        data-prefix="fas"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Button has a valid snapshot 1`] = `
<button
  className="fake-class"
  onClick={[Function]}
  type="button"
>
  Button Text
</button>
`;

exports[`Search has a valid snapshot 1`] = `
<form>
  Search
  <input
    onChange={[Function]}
    type="text"
  />
  <button
    className=""
    onClick={[Function]}
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`Table has a valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span
      className="width-40"
    >
      <div
        className="flex flex-justify-start"
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Title
        </button>
        <div
          className="padding-horizontal-1"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-angle-down fa-w-10 "
            data-icon="angle-down"
            data-prefix="fas"
            role="img"
            style={Object {}}
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
      </div>
    </span>
    <span
      className="width-30"
    >
      <div
        className="flex flex-justify-start"
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Author
        </button>
        <div
          className="padding-horizontal-1"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-angle-down fa-w-10 "
            data-icon="angle-down"
            data-prefix="fas"
            role="img"
            style={Object {}}
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
      </div>
    </span>
    <span
      className="width-10"
    >
      <div
        className="flex flex-justify-start"
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Comments
        </button>
        <div
          className="padding-horizontal-1"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-angle-down fa-w-10 "
            data-icon="angle-down"
            data-prefix="fas"
            role="img"
            style={Object {}}
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
      </div>
    </span>
    <span
      className="width-10"
    >
      <div
        className="flex flex-justify-start"
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Points
        </button>
        <div
          className="padding-horizontal-1"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-angle-down fa-w-10 "
            data-icon="angle-down"
            data-prefix="fas"
            role="img"
            style={Object {}}
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
      </div>
    </span>
    <span
      className="width-10"
    />
  </div>
  <div
    className="table-row"
  >
    <span
      className="width-40"
    >
      <a>
        1
      </a>
    </span>
    <span
      className="width-30"
    >
      1
    </span>
    <span
      className="width-10"
    >
      1
    </span>
    <span
      className="width-10"
    >
      2
    </span>
    <span>
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      className="width-40"
    >
      <a>
        2
      </a>
    </span>
    <span
      className="width-30"
    >
      2
    </span>
    <span
      className="width-10"
    >
      1
    </span>
    <span
      className="width-10"
    >
      2
    </span>
    <span>
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
